rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/wait.hpp

```diff
@@ -15,15 +15,97 @@
 #ifndef RMW_CONNEXT_SHARED_CPP__WAIT_HPP_
 #define RMW_CONNEXT_SHARED_CPP__WAIT_HPP_
 
+#include <unordered_set>
+
 #include "ndds_include.hpp"
 
 #include "rmw/error_handling.h"
 #include "rmw/impl/cpp/macros.hpp"
 #include "rmw/types.h"
 
 #include "rmw_connext_shared_cpp/condition_error.hpp"
+#include "rmw_connext_shared_cpp/event_converter.hpp"
 #include "rmw_connext_shared_cpp/types.hpp"
 #include "rmw_connext_shared_cpp/visibility_control.h"
+#include "rmw_connext_shared_cpp/connext_static_event_info.hpp"
+
+
+rmw_ret_t __gather_event_conditions(
+  rmw_events_t * events,
+  std::unordered_set<DDS::StatusCondition *> & status_conditions)
+{
+  if (events) {
+    // gather all status conditions and masks
+    for (size_t i = 0; i < events->event_count; ++i) {
+      rmw_event_t * current_event = events->events[i];
+      DDSEntity * dds_entity =
+        static_cast<ConnextCustomEventInfo *>(current_event->data)->get_entity();
+      if (!dds_entity) {
+        RMW_SET_ERROR_MSG("Event handle is null");
+        return RMW_RET_ERROR;
+      }
+      DDS::StatusCondition * status_condition = dds_entity->get_statuscondition();
+      if (!status_condition) {
+        RMW_SET_ERROR_MSG("status condition handle is null");
+        return RMW_RET_ERROR;
+      }
+
+      DDS_StatusMask new_mask = status_condition->get_enabled_statuses() |
+        get_mask_from_rmw(current_event->event_type);
+      status_condition->set_enabled_statuses(new_mask);
+      status_conditions.insert(status_condition);
+    }
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t __handle_active_event_conditions(rmw_events_t * events)
+{
+  // enable a status condition for each event
+  if (events) {
+    for (size_t i = 0; i < events->event_count; ++i) {
+      rmw_event_t * current_event = events->events[i];
+      DDSEntity * dds_entity =
+        static_cast<ConnextCustomEventInfo *>(current_event->data)->get_entity();
+      if (!dds_entity) {
+        RMW_SET_ERROR_MSG("Event handle is null");
+        return RMW_RET_ERROR;
+      }
+      DDS::StatusCondition * status_condition = dds_entity->get_statuscondition();
+      if (!status_condition) {
+        RMW_SET_ERROR_MSG("status condition handle is null");
+        return RMW_RET_ERROR;
+      }
+      // faster to just check if this entity's status condition is triggered
+      // @todo: ross_desmond change all handlers not to search
+      // but to check if they have been triggered
+      // large performance increase when RPC types increase in size
+      bool is_active = false;
+      if (status_condition->get_trigger_value()) {
+        is_active = static_cast<bool>(dds_entity->get_status_changes() &
```

Comment:
With my tests at https://github.com/aws-robotics/rcl/pull/2, it looks like this is always `false`, which means `rcl_wait()` will malfunction. However, `rcl_take_event()` or `ConnextCustomEventInfo::get_status()` seems to be working fine.

Reply:
For some extra information, it always seems that `dds_entity->get_status_changes() == 512`.

Reply:
> it looks like this is always `false`

What is always false? I don't see anything highlighted and assumed you meant 

```
status_condition->get_trigger_value()
```

Reply:
Sorry, I wasn't clear. `status_condition->get_trigger_value()` is actually `true`, but `is_active` still always ends up `false`.

Reply:
This issue was being tracked at https://github.com/aws-robotics/rmw_connext/issues/3 and is now resolved.